cosmos/views.py

```diff
@@ -195,6 +196,108 @@ def terms(request):
+    success_url = None
+
+    # Permissions
+    permission_required = "cosmos.add_photoalbum"
+    raise_exception = True
+
+    def form_valid(self, form):
+        self.object = form.save()
+        for img in self.request.FILES.getlist("photos"):
+            PhotoObject.objects.create(album=self.object, photo=img)
+        return super(PhotoAlbumCreate, self).form_valid(form)
+
+    def get_succes_url(self):
+        return reverse_lazy("resources")
+
+
+class PhotoAlbumDelete(LoginRequiredMixin, PermissionRequiredMixin, DeleteView):
+    model = PhotoAlbum
+    template_name = "photo_album/photo_album_confirm_delete.html"
+    success_url = reverse_lazy("photo_album-list")
+
+    # Permissions
+    permission_required = "cosmos.delete_photoalbum"
+    raise_exception = True
+
+
+class PhotoObjectDelete(LoginRequiredMixin, PermissionRequiredMixin, DeleteView):
+    model = PhotoObject
+    template_name = "photo_album/photo_object_confirm_delete.html"
+
+    # Permissions
+    permission_required = "cosmos.delete_photoobject"
+    rais_exception = True
+
+    def get_success_url(self) -> str:
+        return reverse_lazy("photo_album-view", kwargs={"pk": self.get_object().album.id})
+
+
+def photo_album_add_photo(request, pk):
+    album = get_object_or_404(PhotoAlbum, pk=pk)
+
+    if request.method == "POST":
+        print(request.FILES)
+        for img in request.FILES.getlist("photo"):
+            PhotoObject.objects.create(album=album, photo=img)
+        return redirect(reverse("photo_album-view", kwargs={"pk": album.id}))
+    else:
+        form = PhotoObjectForm()
+
+    return render(request, "photo_album/photo_album_add_photos.html", {"form": form})
+
+
+def photo_album_list(request):
+    newest_album = PhotoAlbum.objects.order_by("-date")[0]
+    if newest_album.date < datetime.date(newest_album.date.year, 8, 1):
+        year = newest_album.date.year - 1
+    else:
+        year = newest_album.date
+    return photo_album_list_year(request, year)
+
+
+def photo_album_list_year(request, year):
+    # Take august 1st as start of new academic year so as to include intro
+    start_academic_year = datetime.datetime(year, 8, 1)
+    end_academic_year = datetime.datetime(year + 1, 7, 31)
+    album_list = PhotoAlbum.objects.filter(date__gte=start_academic_year, date__lte=end_academic_year).order_by("-date")
+
+    # Determine if next or prev buttons should be shown
+    newest_album = PhotoAlbum.objects.order_by("-date")[0]
+    oldest_album = PhotoAlbum.objects.order_by("date")[0]
+    # check prev button
+    if year == oldest_album.date.year and oldest_album.date >= datetime.date(oldest_album.date.year, 8, 1):
+        prev_button = False
+    else:
+        prev_button = True
+
+    # check next button
+    if year + 1 == newest_album.date.year and newest_album.date <= datetime.date(newest_album.date.year, 7, 31):
+        next_button = False
+    else:
+        next_button = True
+    # TODO: check the above logic
```

Comment:
do you want to add tests in the same PR?

Reply:
No, i think it is best to keep tests async for now, until we are all caught up with tests, so that we can move faster with development and get feedback sooner, even if the tests havent been finished yet. And i dont want to write tests rn :P.